### PR TITLE
Do not specify exercise dates in exam exercises

### PR DESCRIPTION
**Pull request by krusche**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.

### Motivation and Context
It is cumbersome and error prone to set all exercise dates (release, due, assessment due) for exams correctly.
Also students with individual working times cannot be reflected which is a problem.
It makes more sense to simply ignore these exercises dates and set them to null to prevent inconsistencies or other issues.
Exams should 

### Description
Instructors cannot set these dates any more when creating or editing exercises for exams.

* Do not allow users to specify exercise dates in exam exercises
* Always use the specified exam dates instead 
  * exam start date --> release date
  * exam end date (individual) --> due date
  * exam result publication date --> assessment due date
* Always set exercise dates in exam exercises to null
* Also allow to specify assessment due date in programming exercises

**Basically refactoring to go through the usage of all the dates in exercise in the current codebase and see if it needs to be adapted to support exam exercises (where all the dates are null)**

Open in the implementation:

- [x] Make sure that students can still submit all exercise types in their (personal) exam working time
- [x] Make sure the assessment works, when the exercise dates are null
- [x] Check that all tests are running properly (some might need to be adapted)

Known issues/open points to do in follow-up:
- allow to specify assessment due date in programming exercises
- fix tutor-exercise-dashboard leaderboard and side-panel
- make sure that tutors cannot assess submissions with no result (first assessment) after the result release date

### Steps for Testing
1. Create an exam with text, modeling, quiz and programming exercises (try out create, edit and import)
2. If possible, double check that release date, due date and assessment due date are set to null in the database
3. Participate as student in the exam and make sure you can submit until the end of the individual working time without any issues
4. Participate as tutor in the assessment of the exam. This should also work without any problems after the end date of the exam (or more precise the latest individual end date of all students). The Tutor Exam Dashboard will be visible after the latest individual end date.
5. After the result release date, a tutor should not be able to override his assessment (403 error should be thrown)

**Please also review the code properly**, there were some major refactoring for the `AssessmentResource` and `AssessmentService`

### CODE REVIEW

**Comment by stefanwaldhauser** [comment]
Some Errors Found in A Recent Test Case:

- Tutors can update programming exercise assessments after the exam publish results date
- Manual programming exercise modal shows that exercise is still open after the exam publish results date
- Tutors can not assess empty modelling exercise submissions (
image)

**Comment by sleiss** [comment]
@stefanwaldhauser Programming exercises should already support that, at least in regards to the locking/unlocking of repositories. We might have to check the online code editor, but the part for the offline IDE should already work fine.

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java [hunk 1]
```diff
@@ -868,7 +868,14 @@ public void setNumberOfOpenMoreFeedbackRequests(Long numberOfOpenMoreFeedbackReq
     }
 
     public boolean isReleased() {
-        ZonedDateTime releaseDate = getReleaseDate();
+        // Exam
+        ZonedDateTime releaseDate;
+        if (this.getExerciseGroup() != null) {
```
Comment (sleiss) [inline src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:873]:
```suggestion
        if (this.hasExerciseGroup()) {
```

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java [hunk 2]
```diff
@@ -33,23 +34,35 @@
...
+        this.examService = examService;
     }
 
     Result submitResult(Result result, Exercise exercise, Double calculatedScore) {
         Double maxScore = exercise.getMaxScore();
-        result.setRatedIfNotExceeded(exercise.getDueDate(), result.getSubmission().getSubmissionDate());
+
+        if (exercise.hasExerciseGroup()) {
+            Exam exam = exercise.getExerciseGroup().getExam();
+            ZonedDateTime latestIndividualExamEndDate = examService.getLatestIndiviudalExamEndDate(exam);
```
Comment (sleiss) [inline src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:59]: I'm not sure, but should this really depend on the last individual end date and not on the one belonging to the StudentExam to which the exam belongs?

Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:59]: I was unsure here too, maybe @krusche can help

File: src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java [hunk 3]
```diff
@@ -535,10 +539,22 @@ else if (participation instanceof ProgrammingExerciseParticipation) {
         log.debug("REST request to create Result for External Submission for Exercise : {}", exerciseId);
 
         Exercise exercise = exerciseService.findOneWithAdditionalElements(exerciseId);
-        if (exercise.getDueDate() == null || ZonedDateTime.now().isBefore(exercise.getDueDate())) {
-            return ResponseEntity.badRequest().headers(HeaderUtil.createFailureAlert(applicationName, true, "result", "externalSubmissionBeforeDueDate",
-                    "External submissions are not supported before the exercise due date.")).build();
+
+        if (exercise.getExerciseGroup() == null) {
```
Comment (sleiss) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ResultResource.java:543]:
```suggestion
        if (exercise.hasExerciseGroup()) {
```

File: src/main/java/de/tum/in/www1/artemis/web/rest/TextSubmissionResource.java [hunk 4]
```diff
@@ -240,6 +244,11 @@ else if (!authorizationCheckService.isAtLeastInstructorForExercise(exercise)) {
         log.debug("REST request to get a text submission without assessment");
         Exercise exercise = exerciseService.findOne(exerciseId);
 
+        Exam exam = null;
+        if (exercise.getExerciseGroup() != null) {
```
Comment (sleiss) [inline src/main/java/de/tum/in/www1/artemis/web/rest/TextSubmissionResource.java:248]:
```suggestion
        if (exercise.hasExerciseGroup()) {
```

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-lifecycle.component.html [hunk 5]
```diff
@@ -20,54 +21,65 @@
...
-                <div
-                    *ngIf="!exercise.assessmentType || exercise.assessmentType === assessmentType.AUTOMATIC"
-                    class="text-danger"
-                    jhiTranslate="artemisApp.programmingExercise.timeline.notSet"
+            <div *ngIf="isExamMode || exercise.dueDate">
+                <jhi-programming-exercise-test-schedule-date-picker
+                    class="test-schedule-date mr-2"
+                    [(ngModel)]="exercise.buildAndTestStudentSubmissionsAfterDueDate"
+                    [startAt]="exercise.dueDate"
+                    [min]="exercise.dueDate"
```
Comment (sleiss) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-lifecycle.component.html:39]: I'm not sure if that's a problem, but what would happen if `isExamMode` is `true` and `exercise.dueDate` is null?
Then this would also be `null` here.

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java (hunk 2 above)
Comment (MaisiKoleni) [reply src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:59]: Optimally, this should use the individual due date. I would consider to add an appropriate method to `ExamService` that returns the individual due date for a student based on: [`StudentExamService.java` Line 112](https://github.com/ls1intum/Artemis/blob/develop/src/main/java/de/tum/in/www1/artemis/service/StudentExamService.java#L112)

File: src/main/webapp/app/exercises/shared/dashboards/tutor/tutor-exercise-dashboard.component.ts [hunk 6]
```diff
@@ -51,6 +52,8 @@ export class TutorExerciseDashboardComponent implements OnInit, AfterViewInit {
     exercise: Exercise;
     modelingExercise: ModelingExercise;
     courseId: number;
+    exam: Exam | null = null;
+    isExamMode = false;
```
Comment (sleiss) [inline src/main/webapp/app/exercises/shared/dashboards/tutor/tutor-exercise-dashboard.component.ts:56]: We are basically already holding this information in the `exam` variable, I'm not sure if it makes sense to make `isExamMode` a method that checks if `exam` is `null`.
But I believe thats a minor issue as the values are only set once if I see it correctly.

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-lifecycle.component.html (hunk 5 above)
Comment (krusche) [reply src/main/webapp/app/exercises/programming/manage/update/programming-exercise-lifecycle.component.html:39]: when I tried it out, it worked. I think startAt and min is ignored when the value is null

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java (hunk 2 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:59]: Theoretically you are right and this should use the individual date.

However, I think the exam correction should NOT start before the last student exam is over.
So tutors should not even be able to navigate into the tutor exam dashboard in this case and then this check is fine as well

**Comment by stefanwaldhauser** [comment]
image
Added an Endpoint to get the latest individual end date also on the client

File: src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java [hunk 7]
```diff
@@ -251,9 +250,10 @@ else if (!authCheckService.isAtLeastInstructorForExercise(exercise)) {
             return badRequest();
         }
 
-        // Tutors cannot start assessing submissions if the exercise due date hasn't been reached yet
-        if (fileUploadExercise.getDueDate() != null && fileUploadExercise.getDueDate().isAfter(ZonedDateTime.now())) {
-            return notFound();
+        // Check if tutors can start assessing the students submission
+        boolean startAssessingSubmissions = this.examSubmissionService.checkIfExerciseDueDateIsReached(fileUploadExercise);
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/web/rest/FileUploadSubmissionResource.java:254]: While this seems to be a correct implementation, it is hard to follow that this method does a useful check for non-exam exercises. See further comment in `ExamSubmissionService`.

File: src/main/java/de/tum/in/www1/artemis/service/ExamSubmissionService.java [hunk 8]
```diff
@@ -125,4 +125,28 @@ private boolean isSubmissionInTime(Exercise exercise, StudentExam studentExam) {
...
     }
+
+    /**
+     * Checks if the current time is after the exercise due date (course exercises)
+     * Checks if the current time is after the latest exam end date (exam exercises)
+     * We need this check in order to see if the assessment of exercises can be started
+     * @param exercise
+     * @return boolean
+     */
+    public boolean checkIfExerciseDueDateIsReached(Exercise exercise) {
```
Comment (jpbernius) [inline src/main/java/de/tum/in/www1/artemis/service/ExamSubmissionService.java:136]: I am wondering if this is the correct place for this method. I would prefer this to be part of a exam independent service.
The way this is currently used makes it hard to understand that this method does have relevance for non-exam exercises if not reading the comment (which I found hard to understand).

File: src/main/webapp/app/course/dashboards/tutor-course-dashboard/tutor-course-dashboard.component.html [hunk 9]
```diff
@@ -16,6 +16,14 @@ <h2 *ngIf="isExamMode">{{ 'artemisApp.tutorCourseDashboard.pageHeaderExam' | tra
             </h2>
         </div>
     </div>
+    <div *ngIf="isExamMode" class="row justify-content-between">
+        <div class="col-8">
+            <ul>
+                <li><span jhiTranslate="artemisApp.exam.endDate"> Exam End Date</span> : {{ exam?.endDate | artemisDate }}</li>
+                <li><span jhiTranslate="artemisApp.exam.publishResultsDate"> Publish Results Date</span> : {{ exam?.publishResultsDate | artemisDate }}</li>
```
Comment (jpbernius) [inline src/main/webapp/app/course/dashboards/tutor-course-dashboard/tutor-course-dashboard.component.html:23]: Spaces on Purpose?
```suggestion
                <li><span jhiTranslate="artemisApp.exam.endDate">Exam End Date</span> : {{ exam?.endDate | artemisDate }}</li>
                <li><span jhiTranslate="artemisApp.exam.publishResultsDate">Publish Results Date</span> : {{ exam?.publishResultsDate | artemisDate }}</li>
```

File: src/main/webapp/app/exercises/programming/assess/manual-result/programming-assessment-manual-result-dialog.component.ts [hunk 10]
```diff
@@ -137,7 +140,15 @@ export class ProgrammingAssessmentManualResultDialogComponent implements OnInit
                 tap(({ body: participation }) => {
                     this.participation = participation! as ProgrammingExerciseStudentParticipation;
                     this.result.participation = this.participation;
-                    this.isOpenForSubmission = this.participation.exercise.dueDate === null || this.participation.exercise.dueDate.isAfter(moment());
+
+                    if (!!this.exercise.exerciseGroup) {
+                        const exam = this.participation.exercise.exerciseGroup!.exam!;
```
Comment (jpbernius) [inline src/main/webapp/app/exercises/programming/assess/manual-result/programming-assessment-manual-result-dialog.component.ts:145]: Not 100% sure, but I think unwrapping `exerciseGroup` should be implicit by if statement?
```suggestion
                        const exam = this.participation.exercise.exerciseGroup.exam!;
```

File: src/main/java/de/tum/in/www1/artemis/service/ExamSubmissionService.java [hunk 11]
```diff
@@ -125,4 +125,28 @@ private boolean isSubmissionInTime(Exercise exercise, StudentExam studentExam) {
...
+        Exam exam = exercise.getExerciseGroup().getExam();
+        // Tutors cannot start assessing submissions if the exercise due date hasn't been reached yet
+        if (exam != null) {
+            ZonedDateTime latestIndividualExamEndDate = this.examService.getLatestIndiviudalExamEndDate(exam);
+            if (latestIndividualExamEndDate != null && latestIndividualExamEndDate.isAfter(ZonedDateTime.now())) {
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/ExamSubmissionService.java:141]: What do you mean @jpbernius ?

File: src/main/java/de/tum/in/www1/artemis/service/ExamSubmissionService.java [hunk 12]
```diff
@@ -125,4 +125,28 @@ private boolean isSubmissionInTime(Exercise exercise, StudentExam studentExam) {
...
+                return false;
+            }
+        }
+        else {
+            if (exercise.getDueDate() != null && exercise.getDueDate().isAfter(ZonedDateTime.now())) {
+                return false;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (stefanwaldhauser) [reply src/main/java/de/tum/in/www1/artemis/service/ExamSubmissionService.java:149]: @jpbernius  I do not understand your first comment: "The current implementation is consistent with the regular assessment on Artemis."

File: src/main/java/de/tum/in/www1/artemis/service/ExamSubmissionService.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (jpbernius) [reply src/main/java/de/tum/in/www1/artemis/service/ExamSubmissionService.java:141]: For non-exam exercises, you can also assess if no end date is specified.

**Comment by stefanwaldhauser** [comment]
image
Assesments will not be updated in exams

File: src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java (hunk 2 above)
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/service/AssessmentService.java:57]: We disabled the tutor exam dashboard button until the last individual exam end date is reached. So this should be fine :)

File: src/main/java/de/tum/in/www1/artemis/service/ExamSubmissionService.java (hunk 8 above)
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/service/ExamSubmissionService.java]: @jpbernius I moved it to the SubmissionService :)

File: src/main/webapp/app/exercises/programming/assess/manual-result/programming-assessment-manual-result-dialog.component.ts (hunk 10 above)
Comment (fde312) [reply src/main/webapp/app/exercises/programming/assess/manual-result/programming-assessment-manual-result-dialog.component.ts:145]: I get a warning with "Object is possibly null". Therefore, I would leave it :)

Comment (jpbernius) [reply src/main/webapp/app/exercises/programming/assess/manual-result/programming-assessment-manual-result-dialog.component.ts:145]: 🆗

**Comment by sleiss** [comment]
I can't override text assessments @fde312 :
image

**Comment by anditurdiu** [comment]
After an extensive test session with 6 developers and testers we concluded that most of the core functionality is working as expected. The only issue we found was in relation to the programming exam exercises. When that is fixed we can approve :)
